File: test/papismoke/basicsmoke.py
#!/usr/bin/env python

import nowapi as na
import now
import db

nump = 100
numc = 50
nums = 25
nume = 1000
numv = 100

def createdb():
    with now.Connection("localhost", "55505", None, None) as c:
        db.createDB(c, "db150")
        return db.addRandomData(c, nump, numc, nums, nume, numv)

def countdb(c, t):
    with c.cursor() as cur:
         cur.setRowFormat(na.listrow)
         cur.execute('select count(*) from %s' % t)
         for row in cur:
             return row[0]

if __name__ == '__main__':
        (products, clients, stores, buys, visits) = createdb()

        pl = len(products)
        cl = len(clients)
        sl = len(stores)
        bl = len(buys)
        vl = len(visits)

	if pl != nump or \
           cl != numc or \
           sl != nums or \
           bl != nume or \
           vl != numv:
           raise db.TestFailed('what we got is not what we demanded...')
        
        print("having %d | %d | %d | %d | %d" % (len(products), len(clients), \
                                                 len(stores), len(buys), len(visits)))

        with na.connect('localhost', '55505', None, None, 'db150') as c:
             n = countdb(c, 'product')
             if n != pl:
                raise db.TestFailed('products (%d) differs from %d' % (n, len(products)))
             print('products OK')
             n = countdb(c, 'client')
             if n != cl:
                raise db.TestFailed('clients  (%d) differs from %d' % (n, len(products)))
             print('clients  OK')
             n = countdb(c, 'store')
             if n != sl:
                raise db.TestFailed('stores   (%d) differs from %d' % (n, len(products)))
             print('stores   OK')
             n = countdb(c, 'buys')
             if n != bl:
                raise db.TestFailed('buys     (%d) differs from %d' % (n, len(products)))
             print('buys     OK')
             n = countdb(c, 'visits')
             if n != vl:
                raise db.TestFailed('visits   (%d) differs from %d' % (n, len(products)))
             print('visits   OK')

        print("PASSED")       

File: test/luasmoke/basicsmoke.lua
#!/usr/bin/env lua

now = require('now')
db = require('db')

local rc, con = now.connect('localhost', '55505', nil, nil)
if rc ~= now.OK then
   error("cannot connect: " .. rc .. " (" .. con .. ")")
end

db.createDB(con, 'db250')
db.addRandomData(con, {products=100,
                       clients = 50,
                       stores = 25,
                       buys = 1000,
                       visits = 150})

local function fastcount(tab)
  local stmt = string.format("select count(*) from %s", tab)
  for row in con.rows(stmt) do
      print(string.format("% 8s: % 5d", tab, row.field(0)))
      return row.field(0)
  end
end
local function slowcount(tab)
  local stmt = string.format("select * from %s", tab)
  local cnt=0
  for row in con.rows(stmt) do cnt = cnt + 1 end
  return cnt
end

-- count
if fastcount("client") ~= slowcount("client") then
   error("client count differs")
end
if fastcount("product") ~= slowcount("product") then
   error("product count differs")
end
if fastcount("store") ~= slowcount("store") then
   error("store  count differs")
end
if fastcount("buys") ~= slowcount("buys") then
   error("buys   count differs")
end
if fastcount("visits") ~= slowcount("visits") then
   error("visits count differs")
end

-- time
local function weekday(n)
  if     n == 1 then return 'Sun'
  elseif n == 2 then return 'Mon'
  elseif n == 3 then return 'Tue'
  elseif n == 4 then return 'Wed'
  elseif n == 5 then return 'Thu'
  elseif n == 6 then return 'Fri'
  else               return 'Sat'
  end
end

local rc, n = con.getnow()
if rc ~= now.OK then
   error("cannot get now: " .. rc .. " (" .. n .. ")")
end
local rc, t = con.fromnow(n)
if rc ~= now.OK then
   error("cannot convert from now: " .. rc .. " (" .. t .. ")")
end

local f = string.format('%04d-%02d-%02dT%02d:%02d:%02d.%d',
                           t.year, t.month, t.day, 
                           t.hour, t.min, t.sec,
                           math.floor(t.nano/1000000))

print(f .. " was a " .. weekday(t.wday) .. ", " .. t.yday .. ". day of the year")

print(now.timeformat(t))
print(now.dateformat(t))

local rc, m = con.tonow(t)
if rc ~= now.OK then
   error("cannot convert to now: " .. rc .. " (" .. m .. ")")
end

print(string.format("%d ?= %d", n, m))

if n ~= m then
   error("tonow(fromnow(n)) is not n")
end

